File: src/main/java/gui/FXMLController.java
package gui;

import data.Line;
import de.gsi.chart.XYChart;
import de.gsi.dataset.spi.DoubleDataSet;
import gui.component.inputTabs.ParametersTab;
import gui.component.inputTabs.*;
import gui.component.transformTabs.NormalizeTab;
import gui.component.transformTabs.ShiftTab;
import gui.component.transformTabs.SpikesTab;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.control.ComboBox;
import javafx.scene.control.TabPane;

import java.net.URL;
import java.util.ArrayList;
import java.util.ResourceBundle;

public class FXMLController implements Initializable {
    @FXML
    private XYChart firstChart;
    @FXML
    private XYChart secondChart;
    @FXML
    private XYChart thirdChart;
    @FXML
    private XYChart fourthChart;
    @FXML
    private TabPane parameterTabs;
    @FXML
    private TabPane transformTabs;
    @FXML
    private ComboBox<Integer> generateToCombo;
    @FXML
    private ComboBox<Integer> transformFromCombo;
    @FXML
    private ComboBox<Integer> transformToCombo;

    private Line[] lines;
    private XYChart[] charts;

    private ArrayList<ParametersTab> inputParameters;
    private ArrayList<ParametersTab> transformParameters;

    private void addPlugins(XYChart... charts) {
        for (XYChart c: charts) {
            c.getXAxis().setLabel("x");
            c.getYAxis().setLabel("y");
            //TODO: Fix broken plugins
        }
    }
    private void addToInputParameters(final ParametersTab comp) {
        inputParameters.add(comp);
        comp.setOnApplyClicked((e)->{
            final var dataSet = new DoubleDataSet("Вход");
            final Integer selectedIdx = generateToCombo.getSelectionModel()
                    .getSelectedItem();
            var chart = charts[selectedIdx];
            chart.getDatasets().clear();
            chart.getDatasets().add(dataSet);
            lines[selectedIdx] = comp.generateResult();
            lines[selectedIdx].addToDataset(dataSet);

            if (!transformFromCombo.getItems().contains(selectedIdx)){
                transformFromCombo.getItems().add(selectedIdx);
            }
        });
    }
    private void addToTransformParameters(final ParametersTab comp) {
        transformParameters.add(comp);
        comp.setOnApplyClicked((e)->{
            final var dataSet = new DoubleDataSet("Преобразовано");
            final Integer selectedFromIdx = transformFromCombo.getSelectionModel()
                    .getSelectedItem();
            final Integer selectedToIdx = transformToCombo.getSelectionModel()
                    .getSelectedItem();
            var chart = charts[selectedToIdx];
            chart.getDatasets().clear();
            chart.getDatasets().add(dataSet);
            comp.setInputLine(lines[selectedFromIdx]);
            lines[selectedToIdx] = comp.generateResult();
            lines[selectedToIdx].addToDataset(dataSet);
        });
    }
    @Override
    public void initialize(URL url, ResourceBundle rb) {
        lines = new Line[4];
        charts = new XYChart[4];
        charts[0] = firstChart;
        charts[1] = secondChart;
        charts[2] = thirdChart;
        charts[3] = fourthChart;
        inputParameters = new ArrayList<>();
        addToInputParameters(new LinearParametersTab());
        addToInputParameters(new ExponentialParametersTab());
        addToInputParameters(new PiecewiseParametersTab());
        addToInputParameters(new RandomParametersTab());
        addToInputParameters(new MyRandomParametersTab());
        parameterTabs.getTabs().addAll(inputParameters);
        generateToCombo.getItems().addAll(0,1);
        transformToCombo.valueProperty().addListener((e, oldVal, newVal)->{
            if (newVal != null && newVal.equals(oldVal)) return;
            transformTabs.setDisable(transformFromCombo.getValue() == null);
        });
        transformToCombo.getItems().addAll(2,3);
        transformFromCombo.valueProperty().addListener((e, oldVal, newVal)->{
            if (newVal != null && newVal.equals(oldVal)) return;
            transformTabs.setDisable(transformToCombo.getValue() == null);
        });

        transformParameters = new ArrayList<>();
        addToTransformParameters(new NormalizeTab());
        addToTransformParameters(new ShiftTab());
        addToTransformParameters(new SpikesTab());
        transformTabs.getTabs().addAll(transformParameters);

        addPlugins(firstChart,secondChart,thirdChart,fourthChart);
    }
}

File: src/main/java/gui/FXMLController.kt
package gui

import core.Line
import de.gsi.chart.XYChart
import gui.component.inputTabs.*
import gui.component.transformTabs.NormalizeTab
import gui.component.transformTabs.ShiftTab
import gui.component.transformTabs.SpikesTab
import gui.component.transformTabs.StationarityTab
import infrastructure.DataSetTransforms.dataSetSingle
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.control.ComboBox
import javafx.scene.control.TabPane
import java.net.URL
import java.util.*

class FXMLController : Initializable {
    @FXML
    private lateinit var firstChart: XYChart
    @FXML
    private lateinit var secondChart: XYChart
    @FXML
    private lateinit var thirdChart: XYChart
    @FXML
    private lateinit var fourthChart: XYChart
    @FXML
    private lateinit var inputsTabs: TabPane
    @FXML
    private lateinit var transformTabs: TabPane
    @FXML
    private lateinit var generateToCombo: ComboBox<String>
    @FXML
    private lateinit var transformFromCombo: ComboBox<String>
    @FXML
    private lateinit var transformToCombo: ComboBox<String>

    private var lines = arrayOf(emptyList<Line>(), emptyList(), emptyList(), emptyList())
    private lateinit var charts: Array<XYChart>

    private val inputParameters = listOf(LinearParametersTab(),
            ExponentParametersTab(),
            PiecewiseParametersTab(),
            MyRandomParametersTab(),
            RandomParametersTab()
    )
    private val transformParameters = listOf(NormalizeTab(),
            ShiftTab(),
            SpikesTab(),
            StationarityTab()
    )

    private fun addPlugins(vararg charts: XYChart) {
        for (c in charts) {
            c.xAxis.label = "x"
            c.yAxis.label = "y"
        }
    }

    private fun handleInputsTabs() {
        inputParameters.forEach { comp ->
            comp.onApplyClicked = EventHandler {
                val selectedIdx = generateToCombo.selectionModel
                        .selectedItem.toInt()
                val chart = charts[selectedIdx]
                lines[selectedIdx] = comp.generateResult()
                addToDataset(lines[selectedIdx], chart)

                if (!transformFromCombo.items.contains(selectedIdx.toString())) {
                    transformFromCombo.items.add(selectedIdx.toString())
                }
            }
        }
        inputsTabs.tabs.addAll(inputParameters)
        generateToCombo.items.addAll("0", "1")
        generateToCombo.selectionModel.select(0)
    }

    private fun addToDataset(lines: List<Line>, chart: XYChart) {
        chart.datasets.clear()
        for ((i, l) in lines.withIndex()) {
            chart.datasets.add(dataSetSingle(i.toString(), l).first())
        }
    }

    private fun handleTransformTabs() {
        for (comp in transformParameters) {
            comp.onApplyClicked = EventHandler {
                val selectedFromIdx = transformFromCombo.selectionModel
                        .selectedItem.toInt()
                val selectedToIdx = transformToCombo.selectionModel
                        .selectedItem.toInt()
                comp.inputLine = lines[selectedFromIdx][0]
                val chart = charts[selectedToIdx]
                lines[selectedToIdx] = comp.generateResult()
                addToDataset(lines[selectedToIdx], chart)
            }
        }
        transformTabs.tabs.addAll(transformParameters)
        transformToCombo.items.addAll("2", "3")
        transformToCombo.selectionModel.select(0)
        transformTabs.disableProperty().bind(transformFromCombo.selectionModel
                .selectedItemProperty().isNull)
    }

    override fun initialize(url: URL, rb: ResourceBundle?) {
        charts = arrayOf(firstChart, secondChart, thirdChart, fourthChart)
        handleInputsTabs()
        handleTransformTabs()
        addPlugins(firstChart, secondChart, thirdChart, fourthChart)
    }
}